File: todo-mobile/mobile/test/todo_api_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:todo_mobile/services/todo_api.dart';

void main() {
  group('TodoApi', () {
    test('fetchTodos parses the list', () async {
      final api = TodoApi(
        baseUrl: 'http://test',
        client: MockClient((req) async {
          expect(req.method, 'GET');
          expect(req.url.path, '/api/todos');
          return http.Response(
            jsonEncode([
              {'id': 1, 'title': 'Buy milk', 'completed': false},
              {'id': 2, 'title': 'Ship app', 'completed': true},
            ]),
            200,
            headers: {'content-type': 'application/json'},
          );
        }),
      );

      final todos = await api.fetchTodos();
      expect(todos, hasLength(2));
      expect(todos.first.title, 'Buy milk');
      expect(todos.last.completed, isTrue);
    });

    test('addTodo posts the title and parses the created todo', () async {
      final api = TodoApi(
        baseUrl: 'http://test',
        client: MockClient((req) async {
          expect(req.method, 'POST');
          expect(jsonDecode(req.body), {'title': 'Buy milk'});
          return http.Response(
            jsonEncode({'id': 1, 'title': 'Buy milk', 'completed': false}),
            201,
            headers: {'content-type': 'application/json'},
          );
        }),
      );

      final todo = await api.addTodo('Buy milk');
      expect(todo.id, 1);
      expect(todo.completed, isFalse);
    });

    test('addTodo surfaces the server error message', () async {
      final api = TodoApi(
        baseUrl: 'http://test',
        client: MockClient(
          (req) async => http.Response(
            jsonEncode({'error': 'title is required'}),
            400,
            headers: {'content-type': 'application/json'},
          ),
        ),
      );

      expect(
        () => api.addTodo('   '),
        throwsA(
          isA<TodoApiException>()
              .having((e) => e.statusCode, 'statusCode', 400)
              .having((e) => e.message, 'message', 'title is required'),
        ),
      );
    });

    test('setCompleted patches the todo', () async {
      final api = TodoApi(
        baseUrl: 'http://test',
        client: MockClient((req) async {
          expect(req.method, 'PATCH');
          expect(req.url.path, '/api/todos/7');
          expect(jsonDecode(req.body), {'completed': true});
          return http.Response(
            jsonEncode({'id': 7, 'title': 'x', 'completed': true}),
            200,
            headers: {'content-type': 'application/json'},
          );
        }),
      );

      final todo = await api.setCompleted(7, true);
      expect(todo.completed, isTrue);
    });

    test('deleteTodo throws on 404', () async {
      final api = TodoApi(
        baseUrl: 'http://test',
        client: MockClient(
          (req) async => http.Response(
            jsonEncode({'error': 'not found'}),
            404,
            headers: {'content-type': 'application/json'},
          ),
        ),
      );

      expect(
        () => api.deleteTodo(99),
        throwsA(
          isA<TodoApiException>()
              .having((e) => e.statusCode, 'statusCode', 404),
        ),
      );
    });
  });
}

File: todo-mobile/android-kotlin/app/src/test/kotlin/co/thehindu/todonative/TodoApiTest.kt
package co.thehindu.todonative

import kotlinx.coroutines.test.runTest
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.json.JSONObject
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Before
import org.junit.Test

class TodoApiTest {

    private lateinit var server: MockWebServer
    private lateinit var api: TodoApi

    @Before
    fun setUp() {
        server = MockWebServer()
        server.start()
        api = TodoApi(server.url("/").toString().removeSuffix("/"))
    }

    @After
    fun tearDown() {
        server.shutdown()
    }

    private fun json(body: String, code: Int = 200): MockResponse =
        MockResponse().setResponseCode(code)
            .setHeader("Content-Type", "application/json")
            .setBody(body)

    @Test
    fun `fetchTodos parses the list`() = runTest {
        server.enqueue(
            json("""[{"id":1,"title":"Buy milk","completed":false},{"id":2,"title":"Ship app","completed":true}]"""),
        )

        val todos = api.fetchTodos()

        assertEquals(2, todos.size)
        assertEquals("Buy milk", todos[0].title)
        assertTrue(todos[1].completed)
        assertEquals("/api/todos", server.takeRequest().path)
    }

    @Test
    fun `addTodo posts the title and parses the created todo`() = runTest {
        server.enqueue(json("""{"id":1,"title":"Buy milk","completed":false}""", 201))

        val todo = api.addTodo("Buy milk")

        assertEquals(1, todo.id)
        assertFalse(todo.completed)
        val request = server.takeRequest()
        assertEquals("POST", request.method)
        assertEquals("Buy milk", JSONObject(request.body.readUtf8()).getString("title"))
    }

    @Test
    fun `addTodo surfaces the server error message`() = runTest {
        server.enqueue(json("""{"error":"title is required"}""", 400))

        try {
            api.addTodo("   ")
            fail("expected TodoApiException")
        } catch (e: TodoApiException) {
            assertEquals(400, e.statusCode)
            assertEquals("title is required", e.message)
        }
    }

    @Test
    fun `setCompleted patches the todo`() = runTest {
        server.enqueue(json("""{"id":7,"title":"x","completed":true}"""))

        val todo = api.setCompleted(7, true)

        assertTrue(todo.completed)
        val request = server.takeRequest()
        assertEquals("PATCH", request.method)
        assertEquals("/api/todos/7", request.path)
        assertTrue(JSONObject(request.body.readUtf8()).getBoolean("completed"))
    }

    @Test
    fun `deleteTodo succeeds on 204`() = runTest {
        server.enqueue(MockResponse().setResponseCode(204))

        api.deleteTodo(3)

        val request = server.takeRequest()
        assertEquals("DELETE", request.method)
        assertEquals("/api/todos/3", request.path)
    }

    @Test
    fun `deleteTodo throws on 404`() = runTest {
        server.enqueue(json("""{"error":"not found"}""", 404))

        try {
            api.deleteTodo(99)
            fail("expected TodoApiException")
        } catch (e: TodoApiException) {
            assertEquals(404, e.statusCode)
        }
    }
}
